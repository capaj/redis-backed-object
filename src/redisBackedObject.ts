import Redis from 'ioredis'
import { cloneDeep, debounce, isSymbol } from 'lodash-es'
import mitt, { Emitter } from 'mitt'
/**
 * RedisBackedObject is a class that allows you to create an object that is backed by Redis. This can be useful for caching state in long running processes.
 * Usually you manually save state to Redis, but with RedisBackedObject you can just modify the object and it will automatically save to Redis.
 */
export class RedisBackedObject<T extends object> {
  private redis: Redis
  private key: string
  private initialData: T
  proxy: T
  saveInterval: number
  initPromise: Promise<void>
  emitter: Emitter<
    Record<
      string,
      {
        path: string[]
        value: T
      }
    >
  >

  constructor(
    {
      redis,
      key,
      saveInterval = 1000
    }: {
      redis: Redis
      key: string
      saveInterval?: number
    },
    initialData: T
  ) {
    this.emitter = mitt()
    this.redis = redis
    this.saveInterval = saveInterval
    this.key = key
    this.initialData = cloneDeep(initialData)
    this.proxy = this.deepProxy(initialData, null)

    this.initPromise = this.init()
  }

  private async init() {
    const storedData = await this.redis.get(this.key)
    if (!storedData) {
      return
    }
    const parsed = JSON.parse(storedData)
    if (this.proxy) {
      Object.assign(this.proxy, parsed)
    }
  }

  private deepProxy = (data: any, parentPath: string[]): any => {
    if (typeof data !== 'object' || data === null) {
      return data // Only proxy objects and arrays
    }

    for (const key in data) {
      data[key] = this.deepProxy(
        data[key],
        Array.isArray(parentPath) ? [...parentPath, key] : [key]
      )
    }

    return new Proxy(data, {
      set: (target, property, value) => {
        const propertyPath = isSymbol(property) ? property.toString() : property
        target[property] =
          typeof value === 'object'
            ? this.deepProxy(
                value,
                Array.isArray(parentPath)
                  ? [...parentPath, propertyPath]
                  : [propertyPath]
              )
            : value
        this.emitter.emit('set', {
          path: parentPath ? [...parentPath, propertyPath] : [propertyPath],
          value: this.proxy
        })
        this.debouncedSave()
        return true
      },
      deleteProperty: (target, property) => {
        this.emitter.emit('delete', {
          path: parentPath,
          value: this.proxy
        })
        this.debouncedSave()
        return delete target[property]
      }
    })
  }

  saveToRedis = async () => {
    await this.redis.set(this.key, JSON.stringify(this.proxy))
    this.emitter.emit('save', {
      path: [],
      value: this.proxy
    })
  }

  // @ts-expect-error works ok
  debouncedSave = debounce(this.saveToRedis, this.saveInterval, {
    leading: false,
    trailing: true // Save after the last change in the interval, other way around we would not save the last change into redis in case many changes happen in the interval
  })

  public reset() {
    Object.assign(this.proxy, this.initialData)
    this.emitter.emit('reset', {
      path: [],
      value: this.initialData
    })
  }
}
