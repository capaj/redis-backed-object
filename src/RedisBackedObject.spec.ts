import { beforeEach, describe, expect, it } from 'vitest'
import { RedisBackedObject } from './redisBackedObject'
import Redis, { RedisOptions } from 'ioredis'

const dbOrdinal = Number(process.env.VITEST_WORKER_ID) % 16 // by default redis has 16 databases, we need to tweak the config if we need more

export const redisOptions: RedisOptions = {
  host: process.env.REDIS_HOST || 'localhost',
  port: 6379,
  db: process.env.VITEST_WORKER_ID ? dbOrdinal : undefined,
  connectTimeout: 2000,
  retryStrategy: (times: number) => Math.min(times * 50, 2000),
  enableAutoPipelining: true
}

export const redisClient = new Redis(redisOptions)

const sleep = (ms: number) => new Promise((resolve) => setTimeout(resolve, ms))

describe('RedisBackedObject', () => {
  beforeEach(async () => {
    await redisClient.del('rbo:test')
    await redisClient.flushdb()
  })

  it('should save to Redis', async () => {
    const state = new RedisBackedObject(
      {
        redis: redisClient,
        key: 'rbo:test',
        saveInterval: 2
      },
      {
        a: 1
      }
    )
    const proxy = state.proxy
    proxy.a = 2

    await sleep(2)
    const data = await redisClient.get('rbo:test')
    expect(data).toBe('{"a":2}')

    proxy.a = 3
    await sleep(2)
    const data2 = await redisClient.get('rbo:test')
    expect(data2).toBe('{"a":3}')

    // @ts-expect-error
    delete proxy.a
    await sleep(2)
    const data3 = await redisClient.get('rbo:test')
    expect(data3).toBe('{}')

    // @ts-expect-error
    proxy.array = [1, 2, 3]
    await sleep(2)
    const data4 = await redisClient.get('rbo:test')
    expect(data4).toBe('{"array":[1,2,3]}')

    // @ts-expect-error
    proxy.array.push(4)
    await sleep(2)
    const data5 = await redisClient.get('rbo:test')
    expect(data5).toBe('{"array":[1,2,3,4]}')

    // @ts-expect-error
    proxy.array.pop()
    await sleep(2)
    const data5b = await redisClient.get('rbo:test')
    expect(data5b).toBe('{"array":[1,2,3]}')

    // nested objects
    // @ts-expect-error
    proxy.nested = { a: [1] }
    await sleep(2)
    const data6 = await redisClient.get('rbo:test')
    expect(data6).toBe('{"array":[1,2,3],"nested":{"a":[1]}}')
    // @ts-expect-error
    proxy.nested.a.push(2)
    await sleep(2)
    const data7 = await redisClient.get('rbo:test')
    expect(data7).toBe('{"array":[1,2,3],"nested":{"a":[1,2]}}')
  })

  it('should load from Redis', async () => {
    await redisClient.set('rbo:test2', '{"a":4}')

    const state = new RedisBackedObject(
      {
        redis: redisClient,
        key: 'rbo:test2',
        saveInterval: 2
      },
      {
        a: 1
      }
    )

    const proxy = state.proxy
    expect(proxy.a).toBe(1) // before init runs we have the default

    await sleep(65)
    expect(proxy.a).toBe(4)
  })

  it('should reset to initial state', async () => {
    const state = new RedisBackedObject(
      {
        redis: redisClient,
        key: 'rbo:test3',
        saveInterval: 2
      },
      {
        a: 15
      }
    )
    const proxy = state.proxy
    proxy.a = 2
    await sleep(2)
    const data = await redisClient.get('rbo:test3')
    expect(data).toBe('{"a":2}')

    state.reset()
    expect(proxy.a).toBe(15)
    await sleep(2)
    const data2 = await redisClient.get('rbo:test3')
    expect(data2).toBe('{"a":15}')
  })
})
