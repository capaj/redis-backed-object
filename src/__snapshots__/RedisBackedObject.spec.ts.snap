// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`RedisBackedObject > should save to Redis 1`] = `
[
  [
    "set",
    {
      "path": "a",
      "value": {
        "a": 2,
      },
    },
  ],
  [
    "save",
    {
      "path": null,
      "value": {
        "a": 2,
      },
    },
  ],
]
`;

exports[`RedisBackedObject > should save to Redis 2`] = `
[
  [
    "set",
    {
      "path": "a",
      "value": {},
    },
  ],
  [
    "save",
    {
      "path": null,
      "value": {},
    },
  ],
  [
    "set",
    {
      "path": "a",
      "value": {},
    },
  ],
  [
    "save",
    {
      "path": null,
      "value": {},
    },
  ],
  [
    "delete",
    {
      "path": null,
      "value": {},
    },
  ],
  [
    "save",
    {
      "path": null,
      "value": {},
    },
  ],
]
`;

exports[`RedisBackedObject > should save to Redis 3`] = `
[
  [
    "set",
    {
      "path": "a",
      "value": {
        "array": [
          1,
          2,
          3,
          4,
        ],
      },
    },
  ],
  [
    "save",
    {
      "path": null,
      "value": {
        "array": [
          1,
          2,
          3,
          4,
        ],
      },
    },
  ],
  [
    "set",
    {
      "path": "a",
      "value": {
        "array": [
          1,
          2,
          3,
          4,
        ],
      },
    },
  ],
  [
    "save",
    {
      "path": null,
      "value": {
        "array": [
          1,
          2,
          3,
          4,
        ],
      },
    },
  ],
  [
    "delete",
    {
      "path": null,
      "value": {
        "array": [
          1,
          2,
          3,
          4,
        ],
      },
    },
  ],
  [
    "save",
    {
      "path": null,
      "value": {
        "array": [
          1,
          2,
          3,
          4,
        ],
      },
    },
  ],
  [
    "set",
    {
      "path": "array",
      "value": {
        "array": [
          1,
          2,
          3,
          4,
        ],
      },
    },
  ],
  [
    "save",
    {
      "path": null,
      "value": {
        "array": [
          1,
          2,
          3,
          4,
        ],
      },
    },
  ],
  [
    "set",
    {
      "path": [
        "array",
        "3",
      ],
      "value": {
        "array": [
          1,
          2,
          3,
          4,
        ],
      },
    },
  ],
  [
    "set",
    {
      "path": [
        "array",
        "length",
      ],
      "value": {
        "array": [
          1,
          2,
          3,
          4,
        ],
      },
    },
  ],
  [
    "save",
    {
      "path": null,
      "value": {
        "array": [
          1,
          2,
          3,
          4,
        ],
      },
    },
  ],
]
`;

exports[`RedisBackedObject > should save to Redis 4`] = `
[
  [
    "set",
    {
      "path": "a",
      "value": {
        "array": [
          1,
          2,
          3,
        ],
        "nested": {
          "a": [
            1,
            2,
          ],
        },
      },
    },
  ],
  [
    "save",
    {
      "path": null,
      "value": {
        "array": [
          1,
          2,
          3,
        ],
        "nested": {
          "a": [
            1,
            2,
          ],
        },
      },
    },
  ],
  [
    "set",
    {
      "path": "a",
      "value": {
        "array": [
          1,
          2,
          3,
        ],
        "nested": {
          "a": [
            1,
            2,
          ],
        },
      },
    },
  ],
  [
    "save",
    {
      "path": null,
      "value": {
        "array": [
          1,
          2,
          3,
        ],
        "nested": {
          "a": [
            1,
            2,
          ],
        },
      },
    },
  ],
  [
    "delete",
    {
      "path": null,
      "value": {
        "array": [
          1,
          2,
          3,
        ],
        "nested": {
          "a": [
            1,
            2,
          ],
        },
      },
    },
  ],
  [
    "save",
    {
      "path": null,
      "value": {
        "array": [
          1,
          2,
          3,
        ],
        "nested": {
          "a": [
            1,
            2,
          ],
        },
      },
    },
  ],
  [
    "set",
    {
      "path": "array",
      "value": {
        "array": [
          1,
          2,
          3,
        ],
        "nested": {
          "a": [
            1,
            2,
          ],
        },
      },
    },
  ],
  [
    "save",
    {
      "path": null,
      "value": {
        "array": [
          1,
          2,
          3,
        ],
        "nested": {
          "a": [
            1,
            2,
          ],
        },
      },
    },
  ],
  [
    "set",
    {
      "path": [
        "array",
        "3",
      ],
      "value": {
        "array": [
          1,
          2,
          3,
        ],
        "nested": {
          "a": [
            1,
            2,
          ],
        },
      },
    },
  ],
  [
    "set",
    {
      "path": [
        "array",
        "length",
      ],
      "value": {
        "array": [
          1,
          2,
          3,
        ],
        "nested": {
          "a": [
            1,
            2,
          ],
        },
      },
    },
  ],
  [
    "save",
    {
      "path": null,
      "value": {
        "array": [
          1,
          2,
          3,
        ],
        "nested": {
          "a": [
            1,
            2,
          ],
        },
      },
    },
  ],
  [
    "delete",
    {
      "path": [
        "array",
      ],
      "value": {
        "array": [
          1,
          2,
          3,
        ],
        "nested": {
          "a": [
            1,
            2,
          ],
        },
      },
    },
  ],
  [
    "set",
    {
      "path": [
        "array",
        "length",
      ],
      "value": {
        "array": [
          1,
          2,
          3,
        ],
        "nested": {
          "a": [
            1,
            2,
          ],
        },
      },
    },
  ],
  [
    "save",
    {
      "path": null,
      "value": {
        "array": [
          1,
          2,
          3,
        ],
        "nested": {
          "a": [
            1,
            2,
          ],
        },
      },
    },
  ],
  [
    "set",
    {
      "path": "nested",
      "value": {
        "array": [
          1,
          2,
          3,
        ],
        "nested": {
          "a": [
            1,
            2,
          ],
        },
      },
    },
  ],
  [
    "save",
    {
      "path": null,
      "value": {
        "array": [
          1,
          2,
          3,
        ],
        "nested": {
          "a": [
            1,
            2,
          ],
        },
      },
    },
  ],
  [
    "set",
    {
      "path": [
        "nested",
        "a",
        "1",
      ],
      "value": {
        "array": [
          1,
          2,
          3,
        ],
        "nested": {
          "a": [
            1,
            2,
          ],
        },
      },
    },
  ],
  [
    "set",
    {
      "path": [
        "nested",
        "a",
        "length",
      ],
      "value": {
        "array": [
          1,
          2,
          3,
        ],
        "nested": {
          "a": [
            1,
            2,
          ],
        },
      },
    },
  ],
  [
    "save",
    {
      "path": null,
      "value": {
        "array": [
          1,
          2,
          3,
        ],
        "nested": {
          "a": [
            1,
            2,
          ],
        },
      },
    },
  ],
]
`;
